go/vt/tabletserver/tablet_error.go

```diff
@@ -212,28 +213,89 @@ func logError(queryServiceStats *QueryServiceStats) {
 	}
 }
 
-// AddTabletErrorToQueryResult will mutate a QueryResult struct to fill in the Err
-// field with details from the TabletError.
-func AddTabletErrorToQueryResult(err error, reply *mproto.QueryResult) {
+// rpcErrFromTabletError translate a TabletError to an *mproto.RPCError
+func rpcErrFromTabletError(err error) *mproto.RPCError {
 	if err == nil {
-		return
+		return nil
 	}
-	var rpcErr *mproto.RPCError
 	terr, ok := err.(*TabletError)
 	if ok {
-		rpcErr = &mproto.RPCError{
+		return &mproto.RPCError{
 			// Transform TabletError code to VitessError code
 			Code: int64(terr.ErrorType) + vterrors.TabletError,
 			// Make sure the the VitessError message is identical to the TabletError
 			// err, so that downstream consumers will see identical messages no matter
-			// which endpoint they're using.
+			// which server version they're using.
 			Message: terr.Error(),
 		}
-	} else {
-		rpcErr = &mproto.RPCError{
-			Code:    vterrors.UnknownTabletError,
-			Message: err.Error(),
-		}
 	}
-	reply.Err = rpcErr
+
+	// We don't know exactly what the passed in error was
+	return &mproto.RPCError{
+		Code:    vterrors.UnknownTabletError,
+		Message: err.Error(),
+	}
+}
+
+// AddTabletErrorToQueryResult will mutate a QueryResult struct to fill in the Err
+// field with details from the TabletError.
+func AddTabletErrorToQueryResult(err error, reply *mproto.QueryResult) {
```

Comment:
Not exactly true. The main reason that I want these different functions is because I want strong typing, so I need one function per type. (the alternative would be to specify `interface` in the signature, which I'd prefer to avoid)

The reason that I have the nil check in each function is that, if err is nil, I don't want to overwrite the existing reply.Err with nil. The current behavior will ignore reply.Err when err is nil.